File: plans/network/pingpong.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"strings"
	"time"

	"github.com/testground/sdk-go/network"
	"github.com/testground/sdk-go/run"
	"github.com/testground/sdk-go/runtime"
	"github.com/testground/sdk-go/sync"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
)

func pingpong(runenv *runtime.RunEnv, initCtx *run.InitContext) error {
	ctx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
	defer cancel()

	runenv.RecordMessage("before sync.MustBoundClient")
	client := initCtx.SyncClient
	netclient := initCtx.NetClient

	oldAddrs, err := net.InterfaceAddrs()
	if err != nil {
		return err
	}

	config := &network.Config{
		// Control the "default" network. At the moment, this is the only network.
		Network: "default",

		// Enable this network. Setting this to false will disconnect this test
		// instance from this network. You probably don't want to do that.
		Enable: true,
		Default: network.LinkShape{
			Latency:   100 * time.Millisecond,
			Bandwidth: 1 << 20, // 1Mib
		},
		CallbackState: "network-configured",
		RoutingPolicy: network.DenyAll,
	}

	runenv.RecordMessage("before netclient.MustConfigureNetwork")
	netclient.MustConfigureNetwork(ctx, config)

	seq := client.MustSignalAndWait(ctx, "ip-allocation", runenv.TestInstanceCount)

	// Make sure that the IP addresses don't change unless we request it.
	if newAddrs, err := net.InterfaceAddrs(); err != nil {
		return err
	} else if !sameAddrs(oldAddrs, newAddrs) {
		return fmt.Errorf("interfaces changed")
	}

	ipC := byte((seq >> 8) + 1)
	ipD := byte(seq)

	config.IPv4 = runenv.TestSubnet

	var newIp = append(config.IPv4.IP[0:2:2], ipC, ipD)

	runenv.RecordMessage("I am %d, and my desired IP is %s\n", seq, newIp)

	config.IPv4.IP = newIp
	config.IPv4.Mask = []byte{255, 255, 255, 0}
	config.CallbackState = "ip-changed"

	var (
		listener *net.TCPListener
		conn     *net.TCPConn
	)

	if seq == 1 {
		listener, err = net.ListenTCP("tcp4", &net.TCPAddr{Port: 1234})
		if err != nil {
			return err
		}
		defer listener.Close()
	}

	runenv.RecordMessage("before reconfiguring network")
	netclient.MustConfigureNetwork(ctx, config)

	podTargets, err := getTargetPods(runenv.TestRun)
	// podTarget, err := getTargetIp(TEST_STRING)

	if err != nil {
		return err
	}
	fmt.Println("Attempting to connect to ", podTargets)

	switch seq {
	case 1:
		fmt.Println("This container is listening!")
		conn, err = listener.AcceptTCP()
	case 2:
		connected := false
		for _, podTarget := range podTargets {
			var targetIp = net.ParseIP(podTarget)
			fmt.Printf("Attempting to dial %s\n", targetIp)
			conn, err = net.DialTCP("tcp4", nil, &net.TCPAddr{
				IP:   targetIp,
				Port: 1234,
			})
			if err != nil {
				connected = true
				break
			} else {
				fmt.Printf("Error connecting to %s, continuing \n", targetIp)
			}
		}
		if !connected {
			fmt.Printf("Could not connect to any pod target in list %s \n", podTargets)
			return err
		}
	default:
		return fmt.Errorf("expected at most two test instances")
	}
	if err != nil {
		return err
	}

	defer conn.Close()

	// trying to measure latency here.
	err = conn.SetNoDelay(true)
	if err != nil {
		return err
	}

	pingPong := func(test string, rttMin, rttMax time.Duration) error {
		buf := make([]byte, 1)

		runenv.RecordMessage("waiting until ready")

		// wait till both sides are ready
		_, err = conn.Write([]byte{0})
		if err != nil {
			return err
		}

		_, err = conn.Read(buf)
		if err != nil {
			return err
		}

		start := time.Now()

		// write sequence number.
		runenv.RecordMessage("writing my id")
		_, err = conn.Write([]byte{byte(seq)})
		if err != nil {
			return err
		}

		// pong other sequence number
		runenv.RecordMessage("reading their id")
		_, err = conn.Read(buf)
		if err != nil {
			return err
		}

		runenv.RecordMessage("returning their id")
		_, err = conn.Write(buf)
		if err != nil {
			return err
		}

		runenv.RecordMessage("reading my id")
		// read our sequence number
		_, err = conn.Read(buf)
		if err != nil {
			return err
		}

		runenv.RecordMessage("done")

		// stop
		end := time.Now()

		// check the sequence number.
		if buf[0] != byte(seq) {
			return fmt.Errorf("read unexpected value")
		}

		// check the RTT
		rtt := end.Sub(start)
		if rtt < rttMin || rtt > rttMax {
			return fmt.Errorf("expected an RTT between %s and %s, got %s", rttMin, rttMax, rtt)
		}
		runenv.RecordMessage("ping RTT was %s [%s, %s]", rtt, rttMin, rttMax)

		// Don't reconfigure the network until we're done with the first test.
		state := sync.State("ping-pong-" + test)
		client.MustSignalAndWait(ctx, state, runenv.TestInstanceCount)

		return nil
	}
	err = pingPong("200", 0, 21500*time.Millisecond)
	if err != nil {
		return err
	}

	// config.Default.Latency = 10 * time.Millisecond
	// config.CallbackState = "latency-reduced"
	// netclient.MustConfigureNetwork(ctx, config)

	runenv.RecordMessage("ping pong")
	err = pingPong("10", 0, 3500000*time.Millisecond)
	if err != nil {
		return err
	}

	return nil
}

const TEST_STRING = `[{
    "name": "aws-cni",
    "interface": "eth0",
    "ips": [
        "192.168.58.177"
    ],
    "default": true,
    "dns": {}
},{
    "name": "default/weave",
    "ips": [
        "10.43.128.3"
    ],
    "dns": {}
}]`

type K8sPodNetworkInfo struct {
	Name string   `json:"name"`
	Ips  []string `json:"ips"`
}

func getTargetPods(runId string) ([]string, error) {
	// creates the in-cluster config
	config, err := rest.InClusterConfig()
	if err != nil {
		panic(err.Error())
	}
	// creates the clientset
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		panic(err.Error())
	}
	// get pods in all the namespaces by omitting namespace
	// Or specify namespace to get pods in particular namespace
	pods, err := clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
	if err != nil {
		panic(err.Error())
	}
	fmt.Printf("There are %d pods in the cluster\n", len(pods.Items))

	targetIps := make([]string, 0)

	for _, aPod := range pods.Items {
		// look for the pod with runId in the name and 0 as the suffix - that is the listening pod
		if strings.Contains(aPod.Name, runId) {
			fmt.Printf("Found pod belonging to run: %s", aPod.Name)
			networkAnnKey := "k8s.v1.cni.cncf.io/network-status"
			networkAnnotations := aPod.GetAnnotations()[networkAnnKey]
			fmt.Printf("Network annotations: %s", networkAnnotations)
			if networkAnnotations != "" {
				targetIp, err := getTargetIp(networkAnnotations)
				if err != nil {
					return nil, err
				}
				targetIps = append(targetIps, targetIp)
			}
		} else {
			// fmt.Printf("Skipping pod %s\n", aPod.Name)
		}
	}
	return targetIps, nil
}

func getTargetIp(networkAnnotations string) (string, error) {
	networkBody, err := decodePodInfoString(networkAnnotations)

	if err != nil {
		return "", err
	}

	for _, cni := range networkBody {
		if cni.Name == "default/weave" {
			targetIp := cni.Ips[0]
			return targetIp, nil
		}
	}
	return "", fmt.Errorf("could not get target IP")

}

func decodePodInfoString(payload string) ([]K8sPodNetworkInfo, error) {
	var networkBody []K8sPodNetworkInfo
	err := json.Unmarshal([]byte(payload), &networkBody)
	if err != nil {
		return nil, err
	}

	// we expect a list of network interfaces
	if len(networkBody) == 0 {
		return nil, fmt.Errorf("error decoding network annotations into a list")
	}

	return networkBody, nil
}

func sameAddrs(a, b []net.Addr) bool {
	if len(a) != len(b) {
		return false
	}
	aset := make(map[string]bool, len(a))
	for _, addr := range a {
		aset[addr.String()] = true
	}
	for _, addr := range b {
		if !aset[addr.String()] {
			return false
		}
	}
	return true
}

File: plans/example-browser-node/src/pingpong.js
const os = require('os')
const net = require('net')
const ipaddr = require('ipaddr.js')
const { performance } = require('perf_hooks')
const { sync, network } = require('@testground/sdk')


async function pingpong (runenv, client) {
  runenv.recordMessage('before sync.newBoundClient')

  if (!runenv.testSidecar) {
    throw new Error('this test requires a sidecar.')
  }

  let server, socket

  try {
    const netclient = network.newClient(client, runenv)

    runenv.recordMessage('before netclient.waitNetworkInitialized')
    await netclient.waitNetworkInitialized()

    const oldAddrs = os.networkInterfaces()

    const config = {
      network: 'default',
      enable: true,
      default: {
        latency: 100 * 1000 * 1000, // 100ms in nanoseconds
        bandwidth: 1 << 20 // 1 Mib
      },
      callbackState: 'network-configured',
      routingPolicy: network.DENY_ALL
    }

    runenv.recordMessage('before netclient.configureNetwork')
    await netclient.configureNetwork(config)

    const seq = await client.signalAndWait('ip-allocation', runenv.testInstanceCount)

    // Make sure that the IP addresses don't change unless we request it.
    const newAddrs = os.networkInterfaces()
    if (!sameAddrs(oldAddrs, newAddrs)) {
      throw new Error('interfaces changed')
    }

    runenv.recordMessage(`I am ${seq}`)

    const ip = [...runenv.testSubnet[0].octets.slice(0, 2), 1, seq]

    config.IPv4 = `${ip.join('.')}/${runenv.testSubnet[1]}`
    config.callbackState = 'ip-changed'

    if (seq === 1) {
      server = net.createServer()
      server.listen(1234, '0.0.0.0')
    }

    runenv.recordMessage('before reconfiguring network')
    await netclient.configureNetwork(config)

    switch (seq) {
      case 1:
        socket = await new Promise(resolve => {
          server.once('connection', resolve)
        })

        runenv.recordMessage('server is connected!')
        break
      case 2:
        socket = new net.Socket()

        await new Promise(resolve => {
          const raw = ip.slice(0, 3)
          raw.push(1)
          const addr = ipaddr.fromByteArray(raw)
          socket.connect(1234, addr.toString(), resolve)
        })

        runenv.recordMessage('client is connected!')
        break
      default:
        throw new Error('expected at most two test instances')
    }

    socket.setNoDelay(true)

    const write = (data) => new Promise(resolve => {
      socket.write(data, resolve)
    })

    const read = () => new Promise(resolve => {
      socket.once('data', (data) => {
        socket.pause()
        resolve(data)
      })
      socket.resume()
    })

    const pingPong = async (test, rttMin, rttMax) => {
      runenv.recordMessage('waiting until ready')

      // wait until both sides are ready
      await write('0')
      await read()

      const start = performance.now()

      // write sequence number
      runenv.recordMessage('writing my id')
      await write(seq.toString())

      // pong other sequence number
      runenv.recordMessage('reading their id')
      const theirs = await read()

      runenv.recordMessage('returning their id')
      await write(theirs)

      runenv.recordMessage('reading my id')
      const mine = await read()

      runenv.recordMessage('done')

      const end = performance.now()

      // check the sequence number.
      if (seq.toString() !== mine.toString()) {
        throw Error(`read unexpected value, expected ${seq}, received ${mine}`)
      }

      // check the RTT
      const rtt = end - start
      if (rtt < rttMin || rtt > rttMax) {
        throw new Error(`expected an RTT between ${rttMin} and ${rttMax}, got ${rtt}`)
      }

      runenv.recordMessage(`ping RTT was ${rtt} [${rttMin}, ${rttMax}]`)

      // Don't reconfigure the network until we're done with the first test.
      await client.signalAndWait(`ping-pong-${test}`, runenv.testInstanceCount)
    }

    await pingPong('200', 200, 215)

    config.default.latency = 10 * 1000 * 1000 // 10ms in nanoseconds
    config.callbackState = 'latency-reduced'
    await netclient.configureNetwork(config)

    runenv.recordMessage('ping pong')
    await pingPong('10', 20, 35)
  } finally {
    if (server) server.close()
    if (socket) socket.destroy()
  }
}

function sameAddrs (a, b) {
  if (a.length !== b.length) {
    return false
  }

  a = Object.values(a)
    .flat()
    .reduce((acc, curr) => {
      if (!acc.includes(curr.cidr)) {
        acc.push(curr.cidr)
      }
      return acc
    }, [])

  b = Object.values(b).flat()

  for (const { cidr } of b) {
    if (!a.includes(cidr)) {
      return false
    }
  }

  return true
}

module.exports = pingpong
